File: SVM.r
############################Problem 1################################# 
# Load the Dataset
salary_train <- read.csv("C://Users//hp//Desktop//SVM assi//SalaryData_Train.csv", stringsAsFactors = TRUE)
salary_test <- read.csv("C://Users//hp//Desktop//SVM assi//SalaryData_Test.csv", stringsAsFactors = TRUE)

summary(salary_train)
summary(salary_test)


#Training a model on the data
#Begin by training a simple linear SVM
install.packages("kernlab")
library(kernlab)

salary_classifier <- ksvm(Salary ~ ., data = salary_train, kernel = "vanilladot")

#Evaluating model performance
#predictions on testing dataset
salary_predictions <- predict(salary_classifier, salary_test)

table(salary_predictions, salary_test$Salary)
agreement <- salary_predictions == salary_test$Salary
table(agreement)
prop.table(table(agreement))

#Improving model performance
salary_classifier_rbf <- ksvm(Salary ~ ., data = salary_train, kernel = "rbfdot")
salary_predictions_rbf <- predict(salary_classifier_rbf, salary_test)
agreement_rbf <- salary_predictions_rbf == salary_test$Salary
table(agreement_rbf)
prop.table(table(agreement_rbf))

####################################Problem 2################################
# Load the Dataset
library(readr)
forestfires <- read.csv("C://Users//hp//Desktop//SVM assi//forestfires.csv", stringsAsFactors = TRUE)

summary(forestfires)

# Partition Data into train and test data and also droping first 2 columns since already dummies are created in dataset
forestfires_train <- forestfires[1:362, 3:31]
forestfires_test  <- forestfires[363:517, 3:31]

#Training a model on the data
#Begin by training a simple linear SVM
install.packages("kernlab")
library(kernlab)

forestfires_classifier <- ksvm(size_category ~ ., data = forestfires_train, kernel = "vanilladot")

#Evaluating model performance
#predictions on testing dataset
forestfires_predictions <- predict(forestfires_classifier, forestfires_test)

table(forestfires_predictions, forestfires_test$size_category)
agreement <- forestfires_predictions == forestfires_test$size_category
table(agreement)
prop.table(table(agreement))

#Improving model performance
forestfires_classifier_rbf <- ksvm(size_category ~ ., data = forestfires_train, kernel = "rbfdot")
forestfires_predictions_rbf <- predict(forestfires_classifier_rbf, forestfires_test)
agreement_rbf <- forestfires_predictions_rbf == forestfires_test$size_category
table(agreement_rbf)
prop.table(table(agreement_rbf))

#############################END#############################################

File: SVM.py
#################################Problem 1 #############################
#importing required package
import pandas as pd
import numpy as np

#loading dataset
salary_train = pd.read_csv("C:/Users/hp/Desktop/SVM assi/SalaryData_Train.csv")
salary_test = pd.read_csv("C:/Users/hp/Desktop/SVM assi/SalaryData_Test.csv")

#EDA
salary_train.describe()
salary_test.describe()

#label encoding the data
from sklearn import preprocessing
L_enc = preprocessing.LabelEncoder()
salary_train.columns
salary_train['workclass'] = L_enc.fit_transform(salary_train['workclass'])
salary_train['education'] = L_enc.fit_transform(salary_train['education'])
salary_train['maritalstatus'] = L_enc.fit_transform(salary_train['maritalstatus'])
salary_train['occupation'] = L_enc.fit_transform(salary_train['occupation'])
salary_train['relationship'] = L_enc.fit_transform(salary_train['relationship'])
salary_train['race'] = L_enc.fit_transform(salary_train['race'])
salary_train['sex'] = L_enc.fit_transform(salary_train['sex'])
salary_train['native'] = L_enc.fit_transform(salary_train['native'])

salary_test.columns
salary_test['workclass'] = L_enc.fit_transform(salary_test['workclass'])
salary_test['education'] = L_enc.fit_transform(salary_test['education'])
salary_test['maritalstatus'] = L_enc.fit_transform(salary_test['maritalstatus'])
salary_test['occupation'] = L_enc.fit_transform(salary_test['occupation'])
salary_test['relationship'] = L_enc.fit_transform(salary_test['relationship'])
salary_test['race'] = L_enc.fit_transform(salary_test['race'])
salary_test['sex'] = L_enc.fit_transform(salary_test['sex'])
salary_test['native'] = L_enc.fit_transform(salary_test['native'])

#preparing dataset
train_X = salary_train.iloc[:, :13]
train_y = salary_train.iloc[:, 13]
test_X  = salary_test.iloc[:, :13]
test_y  = salary_test.iloc[:, 13]

#building SVM model
# kernel = linear
from sklearn.svm import SVC
model_linear = SVC(kernel = "linear")
#fitting the model
model_linear.fit(train_X, train_y)
#predicting on tests data
pred_test_linear = model_linear.predict(test_X)

#model evaluation accuracy
np.mean(pred_test_linear == test_y)

# kernel = rbf (improving model)
model_rbf = SVC(kernel = "rbf")
#fitting the model
model_rbf.fit(train_X, train_y)
#predicting on test data
pred_test_rbf = model_rbf.predict(test_X)
#model evaluation (accuracy)
np.mean(pred_test_rbf==test_y)

##########################################Problem 2####################
#importing required package
import pandas as pd
import numpy as np

forestfire_data = pd.read_csv("C:/Users/hp/Desktop/SVM assi/forestfires.csv")
forestfire_data.describe()

#droping first 2 columns since dummies area lready created in dataset
forestfire_data = forestfire_data.iloc[:,2:]

from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

train,test = train_test_split(forestfire_data, test_size = 0.20)

#preparing dataset
train_X = train.iloc[:, :28]
train_y = train.iloc[:, 28]
test_X  = test.iloc[:, :28]
test_y  = test.iloc[:, 28]

#EDA
train.describe()
test.describe()

#building SVM model
# kernel = linear
from sklearn.svm import SVC
model_linear = SVC(kernel = "linear")
#fitting the model
model_linear.fit(train_X, train_y)
#predicting on tests data
pred_test_linear = model_linear.predict(test_X)

#model evaluation accuracy
np.mean(pred_test_linear == test_y)

# kernel = rbf (improving model)
model_rbf = SVC(kernel = "rbf")
#fitting the model
model_rbf.fit(train_X, train_y)
#predicting on test data
pred_test_rbf = model_rbf.predict(test_X)
#model evaluation (accuracy)
np.mean(pred_test_rbf==test_y)
######################################END#######################################